File: leetcode/lc_21_MergeTwoSortedList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    
class List:
    def __init__(self, arr:list):
        self.head = None
        self.current = None
        self.newone = None
        first = True
        for i in arr:
            if first:
                self.current = ListNode(i)
                self.head = self.current
                first = False
            else:
                self.newone = ListNode(i)
                self.current.next = self.newone
                self.current = self.current.next
    
    def get_head(self):
        return self.head
    
    def show_list(self):
        current = self.head
        while True:
            print(current.val)
            if current.next:
                current = current.next
            else:
                break

def show_list(input_head):
    current = input_head
    while True:
        print(current.val)
        if current.next:
            current = current.next
        else:
            break


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        cur = dummy = ListNode()
        while list1 and list2:               
            if list1.val < list2.val:
                cur.next = list1
                list1, cur = list1.next, list1
            else:
                cur.next = list2
                list2, cur = list2.next, list2
                
        if list1 or list2:
            cur.next = list1 if list1 else list2
            
        return dummy.next


s = Solution()

l1 = List([1,2,4])
l2 = List([1,3,4])

solved_head = s.mergeTwoLists(l1.get_head(),l2.get_head())

show_list(solved_head)

File: leetcode/lc_21_MergeTwoSortedList.js

const log = console.log

class ListNode{
    constructor(val=0,next=null){
        this.val = val
        this.next = next
    }
}

class List {
    constructor(arr){
        this.head = null
        this.current = null
        let first = true

        arr.forEach(el => {
            if(first){
                this.current = new ListNode(el)
                this.head = this.current
                first = false
            } else {
                let newone = new ListNode(el)
                this.current.next = newone
                this.current = this.current.next
            }
        });
    }

    get_head(){
        return this.head
    }

    show_list(){
        let current = this.head

        while(true){
            
            if(current.next === null){
                log(current.val)
                break
            }else{
                log(current.val)
                current = current.next
            }
        }
    }
}

function show_list(head){
    let current = head

    while(true){
        
        if(current.next === null){
            log(current.val)
            break
        }else{
            log(current.val)
            current = current.next
        }
    }
}

function solution(list1, list2){
    let current = new ListNode()
    let head = new ListNode()
    head = current
    while(list1 && list2){
        if (list1.val < list2.val){
            current.next = list1
            current = list1
            list1 = list1.next
        } else {
            current.next = list2
            current = list2
            list2 = list2.next
        }
    }

    if ( list1 || list2 ){
        current.next = list1? list1:list2
    }
    return head.next
    
}


let l1 = new List([1,2,3,4])

let l2 = new List([2,3,4,5,6,7])

let solved_head = solution(l1.get_head(),l2.get_head())
log(solved_head)
show_list(solved_head)